node/src/components/small_network/task_manager.rs

```diff
@@ -0,0 +1,231 @@
+//! A manager for background tasks.
+//!
+//! Long-running background tasks that need to be shutdown are captured/managed by the task manager,
+//! which allows waiting for all of them to complete before continuing.
+//!
+//! Tasks are passed in a [`ShutdownReceiver`] which signals that a task should end. Upon shutdown,
+//! the task manager will wait for a configurable timeout to shut down, before continuing.
+//!
+//! Example task:
+//!
+//! ```
+//! # use std::time::Duration;
+//! #
+//! # use casper_node::components::small_network::task_manager::TaskManager;
+//! #
+//! let rt = tokio::runtime::Runtime::new().expect("failed to initialize runtime");
+//! let _guard = rt.enter();
+//!
+//! let mut task_manager = TaskManager::new();
+//!
+//! // Spawn 10 tasks that are sleeping for 5 seconds, but responsive to being shut down.
+//! for n in 0..10 {
+//!     task_manager.spawn(|mut shutdown| async move {
+//!         tokio::select! {
+//!             _ = shutdown.wait_for_shutdown() => {
+//!                 // Shutdown received. Simply exit the future.
+//!                 return;
+//!             }
+//!
+//!             _ = tokio::time::sleep(Duration::from_secs(5)) => {
+//!                 println!("5 seconds have passed in task {}", n);
+//!             }
+//!         }
+//!     });
+//! }
+//!
+//! // After we have spawned the tasks, we tell them to shut down and wait for termination.
+//! let tasks_leftover = rt.block_on(
+//!     task_manager.shutdown_and_wait(Duration::from_secs(10))
+//! ).expect("shutdown failed");
+//!
+//! assert_eq!(tasks_leftover, 0);
+//! ```
+//!
+//! # `Drop`
+//!
+//! Dropping the `TaskManager` will cause all tasks to be sent a shutdown notice, but the `Drop`
+//! implementation will **not** wait them to finish.
```

Comment:
Let's have a unit test that proves that `TaskManager` maintains the invariants you specify here.

Reply:
I wanted to write that as well but saw that PR was still in Draft. It's not anymore, let's add tests.

Reply:
Good point. The comment specifying the invariant was added the implementation, but I guess it's fair to test it, now that we are advertising it :)